### ISSUE TITLE
 Hystrix Ignore Exception Global

### ISSUE THREAD

**Issue by hacbq** [Netflix/Hystrix#993]
Hi, can Hystrix next version support Ignore Exception global. Now each methods, I have to clone code
```
@HystrixCommand(ignoreExceptions = { BaseException.class, MissingServletRequestParameterException.class, TypeMismatchException.class })
```
:(

**Comment by mattrjacobs** [comment]
I'm missing some context here - why do you need to add those annotations, and is that a new behavior?

**Comment by shakuzen** [comment]
Here is [some context](https://gitter.im/spring-cloud/spring-cloud?at=564fe0aece18827866f4ff5b). I believe the goal is to be able to configure exceptions to be ignored by all `@HystrixCommand` instead of having to copy-and-paste the above code snippet for every single one. It would be an enhancement request for new behavior - global configuration of ignored exceptions.

That is, unless I've misunderstood the context and desired behavior.

**Comment by mattrjacobs** [comment]
OK, got it.  Thanks for that.  @dmgcodevil, can you take a look at this, since it seems specific to `hystrix-javanica`?

**Comment by dmgcodevil** [comment]
Hi, what if javanica will provide ability to annotate whole class contains Hystrix commands, it's anticipated design, does it work for you  @hacbq, @shakuzen, any suggestions ?

**Comment by mattrjacobs** [comment]
Closing due to inactivity.  Please reopen if there's more to discuss.

**Comment by Geniy00** [comment]
Is there any new related to this question?

I'll describe our problem. We have Spring boot application based on rest services with @ControllerAdvice that handles our exceptions. Usually our internal exceptions just converted to error message dto for customer.
@HystrixCommand annotation wrap any exception and we need to catch HystrixRuntimeException, get cause and handle cause exception. it lead us to duplicate code in @ControllerAdvice controller, and to handle exception in two branches java.lanj.Exception and HystrixRuntimeException.

I think that annotation full class by @HystrixCommand could be good.
If we can turn off exception wrapping by hystrix for trowable via properties file, it will be nice as well.

In my opinion, current solutions can be next:
1. write custom aspect on @HystrixCommand annotated methods that retrows HystrixRuntimeException cause.
2. extend new annotation from @HystrixCommand and use it.

**Comment by mattrjacobs** [comment]
Anyone want to start work on this PR?

**Comment by dmgcodevil** [comment]
@mattrjacobs I've started.

**Comment by dmgcodevil** [comment]
@mattrjacobs I've encountered an issue with one of the patterns, namely:
1. A command throws any custom exception that is not ``HystrixBadRequestException`` hence a fallback (which is a command as well) should be triggered.
2. Fallback throws ``HystrixBadRequestException``
3. A caller gets ``HystrixRuntimeException`` instead of ``HystrixBadRequestException``

Example

```java
public class IgnoreFallbackException {

    private static class FallbackCommand extends HystrixCommand<Void> {
        protected FallbackCommand() {
            super(Setter.withGroupKey(HystrixCommandGroupKey.Factory.asKey("IgnoreFallbackException"))
                    .andCommandKey(HystrixCommandKey.Factory.asKey("FallbackCommand")));
        }

        @Override
        protected Void run() throws Exception {
            throw new HystrixBadRequestException("from fallback");
        }
    }

    private static class SimpleHystrixCommand extends HystrixCommand<Void> {

        protected SimpleHystrixCommand() {
            super(Setter.withGroupKey(HystrixCommandGroupKey.Factory.asKey("IgnoreFallbackException"))
                    .andCommandKey(HystrixCommandKey.Factory.asKey("SimpleHystrixCommand")));
        }

        @Override
        protected Void run() throws Exception {
            throw new RuntimeException("from command");
        }

        @Override
        protected Void getFallback() {
            return new FallbackCommand().execute();
        }
    }

    public static void main(String[] args) {
        new SimpleHystrixCommand().execute();
    }
}
```

Stacktrace

```

Exception in thread "main" com.netflix.hystrix.exception.HystrixRuntimeException: SimpleHystrixCommand failed and fallback failed.
	at com.netflix.hystrix.AbstractCommand$21.call(AbstractCommand.java:793)
	at com.netflix.hystrix.AbstractCommand$21.call(AbstractCommand.java:768)
	at rx.internal.operators.OperatorOnErrorResumeNextViaFunction$4.onError(OperatorOnErrorResumeNextViaFunction.java:139)
	at rx.internal.operators.OperatorDoOnEach$1.onError(OperatorDoOnEach.java:71)
	at rx.internal.operators.OperatorDoOnEach$1.onError(OperatorDoOnEach.java:71)
	at com.netflix.hystrix.AbstractCommand$DeprecatedOnFallbackHookApplication$1.onError(AbstractCommand.java:1448)
	at com.netflix.hystrix.AbstractCommand$FallbackHookApplication$1.onError(AbstractCommand.java:1373)
	at rx.internal.operators.OperatorDoOnEach$1.onError(OperatorDoOnEach.java:71)
	at rx.observers.Subscribers$5.onError(Subscribers.java:224)
	at rx.internal.operators.OnSubscribeThrow.call(OnSubscribeThrow.java:44)
	at rx.internal.operators.OnSubscribeThrow.call(OnSubscribeThrow.java:28)
	at rx.Observable.unsafeSubscribe(Observable.java:8460)
	at rx.internal.operators.OnSubscribeDefer.call(OnSubscribeDefer.java:51)
	at rx.internal.operators.OnSubscribeDefer.call(OnSubscribeDefer.java:35)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:50)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:50)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:50)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:50)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:50)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:50)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:50)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:50)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.Observable.unsafeSubscribe(Observable.java:8460)
	at rx.internal.operators.OperatorOnErrorResumeNextViaFunction$4.onError(OperatorOnErrorResumeNextViaFunction.java:141)
	at rx.internal.operators.OperatorDoOnEach$1.onError(OperatorDoOnEach.java:71)
	at rx.internal.operators.OperatorDoOnEach$1.onError(OperatorDoOnEach.java:71)
	at com.netflix.hystrix.AbstractCommand$HystrixObservableTimeoutOperator$3.onError(AbstractCommand.java:1151)
	at rx.internal.operators.OperatorSubscribeOn$1$1.onError(OperatorSubscribeOn.java:59)
	at rx.observers.Subscribers$5.onError(Subscribers.java:224)
	at rx.internal.operators.OperatorDoOnEach$1.onError(OperatorDoOnEach.java:71)
	at rx.observers.Subscribers$5.onError(Subscribers.java:224)
	at com.netflix.hystrix.AbstractCommand$DeprecatedOnRunHookApplication$1.onError(AbstractCommand.java:1407)
	at com.netflix.hystrix.AbstractCommand$ExecutionHookApplication$1.onError(AbstractCommand.java:1338)
	at rx.observers.Subscribers$5.onError(Subscribers.java:224)
	at rx.internal.operators.OnSubscribeThrow.call(OnSubscribeThrow.java:44)
	at rx.internal.operators.OnSubscribeThrow.call(OnSubscribeThrow.java:28)
	at rx.Observable.unsafeSubscribe(Observable.java:8460)
	at rx.internal.operators.OnSubscribeDefer.call(OnSubscribeDefer.java:51)
	at rx.internal.operators.OnSubscribeDefer.call(OnSubscribeDefer.java:35)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:50)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:50)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.Observable.unsafeSubscribe(Observable.java:8460)
	at rx.internal.operators.OnSubscribeDefer.call(OnSubscribeDefer.java:51)
	at rx.internal.operators.OnSubscribeDefer.call(OnSubscribeDefer.java:35)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:50)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:50)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.Observable.unsafeSubscribe(Observable.java:8460)
	at rx.internal.operators.OperatorSubscribeOn$1.call(OperatorSubscribeOn.java:94)
	at com.netflix.hystrix.strategy.concurrency.HystrixContexSchedulerAction$1.call(HystrixContexSchedulerAction.java:56)
	at com.netflix.hystrix.strategy.concurrency.HystrixContexSchedulerAction$1.call(HystrixContexSchedulerAction.java:47)
	at com.netflix.hystrix.strategy.concurrency.HystrixContexSchedulerAction.call(HystrixContexSchedulerAction.java:69)
	at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.RuntimeException: from command
	at com.netflix.hystrix.contrib.javanica.test.IgnoreFallbackException$SimpleHystrixCommand.run(IgnoreFallbackException.java:35)
	at com.netflix.hystrix.contrib.javanica.test.IgnoreFallbackException$SimpleHystrixCommand.run(IgnoreFallbackException.java:26)
	at com.netflix.hystrix.HystrixCommand$1.call(HystrixCommand.java:293)
	at com.netflix.hystrix.HystrixCommand$1.call(HystrixCommand.java:289)
	at rx.internal.operators.OnSubscribeDefer.call(OnSubscribeDefer.java:46)
	... 23 more
```

No javanica used in example but vanilla hystrix. Is it expected behaviour ?

**Comment by mattrjacobs** [comment]
The semantics of a `HystrixBadRequestException` are to notify the caller that the arguments passed to the command were not appropriate.  The caller gets an exception, and the circuit health is unaffected.  In the case where the run method executes but does not throw `HystrixBadRequestException`, the arguments are valid and I would expect fallback to trigger and the circuit health to get modified.  So I would argue that the only place to throw a `HystrixBadRequestException` is from the `run()` method, and wrapping it from the `getFallback()` is ok.

Thoughts?

**Comment by Geniy00** [comment]
@dmgcodevil 
> It can be not very convinient when class contains 10 public methods and only one should be a hystrix command,
In our project we have additional layer with histrix command, it means that we have separate classes that is used to communicate to another services using hystrix. So, this approach looks good for me, but I'm not sure about other devs.

"backward compatibility"
Up to you

Usual exception handling you can see in Spring class org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler#handleException . We have something similar in our project. So I just need cause, but maybe behavior switcher can be useful for other reasons.

**Comment by dmgcodevil** [comment]
@Geniy00
I've added new annotation ``@DefaultProperties`` that allows to default command properties in addition to ``ignoreExceptions``, you can check it [here](https://github.com/dmgcodevil/Hystrix/blob/584f3622fccb9ea7aedabe2c402e19c208dda0d6/hystrix-contrib/hystrix-javanica/src/main/java/com/netflix/hystrix/contrib/javanica/annotation/DefaultProperties.java), PR still in progress. 

I think it makes sense to always propagate cause by default.

@mattrjacobs 
From this point of view I guess you're right. On the flip side if  ```getFallback``` gets value from another one command that has it's own validation and can throw ```BadRequest``` then maybe it's worth to support this case. Anyway it's not a big problem for now if users aren't requesting this feature.

**Comment by mattrjacobs** [comment]
Closed via #1259

**Comment by cforce** [comment]
How can i use the ignore exception feature inside a FeignClient?
